File: Clojure/problem019.clj
#! /usr/bin/env clojure
; Problem 19 - Project Euler
; http://projecteuler.net/index.php?section=problems&id=19

(defn leap? [year]
  (letfn [(mod-zero? [x d] (zero? (mod x d)))]
    (or (mod-zero? year 400)
        (and (not (mod-zero? year 100))
             (mod-zero? year 4)))))

(defn days-of-year [year]
  (if (leap? year) 366 365))

(defn days-of-months [year]
  (if (leap? year)
    '(31 29 31 30 31 30 31 31 30 31 30 31)
    '(31 28 31 30 31 30 31 31 30 31 30 31)))

(defn amount-days [y m d]
    (->> (range 1901 y)
         (map days-of-year)
         (reduce +)
         (+ (reduce + (take (dec m) (days-of-months y))))
         (+ (dec d))))

(defn sunday? [amount]
  (= (mod amount 7) 5))

(->> (for [y (range 1901 (inc 2000))
           m (range 1 (inc 12))]
       (amount-days y m 1))
     (filter sunday?)
     (count)
     (println))

File: Python/problem019.py
#! /usr/bin/env python3
"""
Problem 19 - Project Euler
http://projecteuler.net/index.php?section=problems&id=019
"""
daysofmonths = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31]


def count_sundays():
    days1901to2000 = []
    for year in range(1901, 2000 + 1):
        if year % 400 == 0 or (year % 100 != 0 and year % 4 == 0):
            dom_leap = daysofmonths[:]
            dom_leap[1] = 29
            days1901to2000.extend(dom_leap)
        else:
            days1901to2000.extend(daysofmonths)

    amountofdays = 0
    sundays = 0
    for days in days1901to2000:
        if amountofdays % 7 == 5:  # 1 Jan 1901 was a Tuesday
            sundays += 1
        amountofdays += days
    return sundays


if __name__ == "__main__":
    print(count_sundays())
